app/components/channel_link/index.js

```diff
@@ -3,26 +3,64 @@
 
 import {connect} from 'react-redux';
 import {bindActionCreators} from 'redux';
+import {createSelector} from 'reselect';
 
-import {getChannelsNameMapInCurrentTeam} from 'mattermost-redux/selectors/entities/channels';
+import {General} from 'mattermost-redux/constants';
+import {
+    getChannelsNameMapInCurrentTeam,
+    getChannelsInCurrentTeam,
+    getMyChannelMemberships,
+} from 'mattermost-redux/selectors/entities/channels';
+import {getCurrentUserId} from 'mattermost-redux/selectors/entities/users';
+import {getCurrentTeamId} from 'mattermost-redux/selectors/entities/teams';
+import {joinChannel, searchChannels} from 'mattermost-redux/actions/channels';
 
 import {handleSelectChannel, setChannelDisplayName} from 'app/actions/views/channel';
 
 import ChannelLink from './channel_link';
 
-function mapStateToProps(state) {
-    return {
-        channelsByName: getChannelsNameMapInCurrentTeam(state),
+function makeGetJoinableChannel() {
+    return createSelector(
+        getChannelsInCurrentTeam,
+        getMyChannelMemberships,
+        (state, props) => props && props.channelMentions,
+        (channels, myMembers, channelMentions) => {
+            const displayname = channelMentions && Object.values(channelMentions)[0].display_name;
```

Comment:
We should use the channel's name here if possible, not the display name since you can have multiple channels with the same display name on a single team